### DISCUSSION TITLE
With <connection>.remote_addrs configured with IPv6 range in CIDR format in swanctl.conf , policy installation failed.

### DISCUSSION THREAD

**Discussion by andy-ning** [strongswan/strongswan#2118] [Q&A]
We have a two IPv6 hosts (named controller-0 and controller-1) cluster running strongswan IPSec. The cluster will expand to more than two hosts eventually. We configured them in host-to-host transport mode. The swanctl.conf are configured as following. We use IP range in CIDR format for remote_addrs, because we want communication with any hosts in the same subnet IPSec protected. 

The immediate issue we are facing is that, when charon is restarted, the logs show policy installation failed. charon_debug.log is attached here [charon_debug.log](https://github.com/strongswan/strongswan/files/14378756/charon_debug.log). And swanctl --list-pol shows no policies, swanctl--list-sa shows no SAs.

Similar configurations work fine for IPv4. (with remote_addrs, local_ts, remote_ts set to 192.168.161.0/24)

**swanctl --version
strongSwan swanctl 5.9.1**

What could be the cause of the issue? Is there any workaround?

- controller-0:

> connections {

        system-nodes {
                reauth_time = 14400
                rekey_time = 3600
                unique = never
                local_addrs = fdff:719a:bf60:2240::3
                remote_addrs = fdff:719a:bf60:2240::/64
                local {
                        auth = pubkey
                        certs = system-ipsec-certificate-controller-0.crt
                }
                remote {
                        id = CN=*
                        auth = pubkey
                        cacerts = system-local-ca.crt
                }
                children {
                        node {
                                mode = transport
                                start_action = trap
                                local_ts = fdff:719a:bf60:2240::/64
                                remote_ts = fdff:719a:bf60:2240::/64
                        }
                }
        }
}

- controller-1:

> connections {

        system-nodes {
                reauth_time = 14400
                rekey_time = 3600
                unique = never
                local_addrs = fdff:719a:bf60:2240::4
                remote_addrs = fdff:719a:bf60:2240::/64
                local {
                        auth = pubkey
                        certs = system-ipsec-certificate-controller-1.crt
                }
                remote {
                        id = CN=*
                        auth = pubkey
                        cacerts = system-local-ca.crt
                }
                children {
                        node {
                                mode = transport
                                start_action = trap
                                local_ts = fdff:719a:bf60:2240::/64
                                remote_ts = fdff:719a:bf60:2240::/64
                        }
                }
        }
}

- Errors in logs:
188 Thu, 2024-02-22, 19:09:09 07[CFG] installing 'node'
189 Thu, 2024-02-22, 19:09:09 07[CFG] configured proposals: ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256, ESP:AES_CBC_128/AES_CBC_192/AES_CB    C_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ
190 Thu, 2024-02-22, 19:09:09 07[KNL] adding policy fdff:719a:bf60:2240::/64 === fdff:719a:bf60:2240::/64 in [priority 334464, refcount 1]
191 Thu, 2024-02-22, 19:09:09 07[KNL] received netlink error: Invalid argument (22)
192 Thu, 2024-02-22, 19:09:09 07[KNL] unable to add policy fdff:719a:bf60:2240::/64 === fdff:719a:bf60:2240::/64 in
193 Thu, 2024-02-22, 19:09:09 07[KNL] adding policy fdff:719a:bf60:2240::/64 === fdff:719a:bf60:2240::/64 out [priority 334464, refcount 1]
194 Thu, 2024-02-22, 19:09:09 07[KNL] received netlink error: Invalid argument (22)
195 Thu, 2024-02-22, 19:09:09 07[KNL] unable to add policy fdff:719a:bf60:2240::/64 === fdff:719a:bf60:2240::/64 out
196 Thu, 2024-02-22, 19:09:09 07[CFG] installing trap failed

- swanctl -L
system-nodes: IKEv1/2, reauthentication every 14400s, rekeying every 3600s
  local:  fdff:719a:bf60:2240::3
  remote: fdff:719a:bf60:2240::/64
  local public key authentication:
    id: CN=ipsec-controller-0
    certs: CN=ipsec-controller-0
  remote public key authentication:
    id: CN=*
    cacerts: CN=starlingx
  node: TRANSPORT, rekeying every 3600s
    local:  fdff:719a:bf60:2240::/64
    remote: fdff:719a:bf60:2240::/64

**Comment by tobiasbrunner** [comment]
> remote_addrs = fdff:719a:bf60:2240::/64

If you configure this, it will result in %any (i.e. an IPv4 address) because it's a subnet that can't be resolved to a single IP address when the trap policy is installed. The only workaround is configuring `remote_addrs = %any6` (or `::`). In such wildcard trap scenarios the hosts you connect to are controlled via `remote_ts` anyway, so the addresses in `local|remote_addrs` don't really matter, other than in regards to the address family.

**Comment by andy-ning** [comment]
Thanks for the rapid reply. 
After changing remote_addrs = %any6, policies are installed successfully and IKE_SAs and CHILD_SAs are established. 

But with this configuration, there is an undesired side effect, charon established IKE_SAs between local addresses, as shown below.

Any way to ensure charon not to establish IKE_SAs within the host between local addresses? 

**system-nodes: #6, ESTABLISHED, IKEv2, 5b886d90bc729998_i 8e93e6bc941653f3_r***   ===> undesired, between local addresses
  local  'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::3[500]
  remote 'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::3[500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_AES128_XCBC/MODP_3072
  established 439s ago, rekeying in 3126s, reauth in 13871s

**system-nodes: #5, ESTABLISHED, IKEv2, **5b886d90bc729998_i\*** 8e93e6bc941653f3_r**  ====> undesired, between local addresses
  local  'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::3[500]
  remote 'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::3[500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_AES128_XCBC/MODP_3072
  established 439s ago, rekeying in 2386s, reauth in 12431s

**system-nodes: #3, ESTABLISHED, IKEv2, 2e672e02ceeab955_i d2f41585c95c62c8_r***
  local  'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::3[500]
  remote 'CN=ipsec-controller-1' @ fdff:719a:bf60:2240::4[500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_AES128_XCBC/MODP_3072
  established 686s ago, rekeying in 1612s, reauth in 12572s

**system-nodes: #2, ESTABLISHED, IKEv2, dc1d3f7350f93b68_i\* 0c5f46006346c8df_r        ====> this is desired.**
  local  'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::3[500]
  remote 'CN=ipsec-controller-1' @ fdff:719a:bf60:2240::4[500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_AES128_XCBC/MODP_3072
  established 687s ago, rekeying in 2668s, reauth in 12174s
  node: #2, reqid 1, INSTALLED, TRANSPORT, ESP:AES_GCM_16-128
    installed 687s ago, rekeying in 2611s, expires in 3273s
    in  c9e77ca1, 105357646 bytes, 89480 packets,     2s ago
    out cfa62814, 1669513 bytes, 38399 packets,     2s ago
    local  fdff:719a:bf60:2240::3/128
    remote fdff:719a:bf60:2240::4/128

- ip xfrm policy

src fdff:719a:bf60:2240::3/128 dst fdff:719a:bf60:2240::4/128
        dir out priority 268927 ptype main
        tmpl src :: dst ::
                proto esp spi 0xcfa62814 reqid 1 mode transport

src fdff:719a:bf60:2240::4/128 dst fdff:719a:bf60:2240::3/128
        dir in priority 268927 ptype main
        tmpl src :: dst ::
                proto esp reqid 1 mode transport

src fdff:719a:bf60:2240::/64 dst fdff:719a:bf60:2240::/64
        dir out priority 334464 ptype main
        tmpl src :: dst ::
                proto esp reqid 1 mode transport

src fdff:719a:bf60:2240::/64 dst fdff:719a:bf60:2240::/64
        dir in priority 334464 ptype main
        tmpl src :: dst ::
                proto esp reqid 1 mode transport

- ip xfrm state

src fdff:719a:bf60:2240::3 dst fdff:719a:bf60:2240::3
        proto esp spi 0x00000000 reqid 1 mode transport
        replay-window 0
        anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
        sel src fdff:719a:bf60:2240::3/128 dst fdff:719a:bf60:2240::3/128 proto ipv6-icmp type 1 code 3

src fdff:719a:bf60:2240::3 dst fdff:719a:bf60:2240::4
        proto esp spi 0xcfa62814 reqid 1 mode transport
        replay-window 0
        aead rfc4106(gcm(aes)) 0xe7d5b5a728a0cf43129439ca490951ea1911d4e4 128
        anti-replay context: seq 0x0, oseq 0xca77, bitmap 0x00000000
        sel src fdff:719a:bf60:2240::3/128 dst fdff:719a:bf60:2240::4/128

src fdff:719a:bf60:2240::4 dst fdff:719a:bf60:2240::3
        proto esp spi 0xc9e77ca1 reqid 1 mode transport
        replay-window 32
        aead rfc4106(gcm(aes)) 0xa7401c1616c26774d13f2dcc1e8163a3d57909b7 128
        anti-replay context: seq 0x1be1f, oseq 0x0, bitmap 0xffffffff
        sel src fdff:719a:bf60:2240::4/128 dst fdff:719a:bf60:2240::3/128

**Reply by tobiasbrunner** [reply]
> Any way to ensure charon not to establish IKE_SAs within the host between local addresses?

Only if you can exclude them somehow via `local|remote_ts` (e.g. reduce `local_ts` to a specific IP). The whole point of configurations like that is to protect traffic between the configured IPs/subnets, which is reflected by the IPsec trap policies that in turn will obviously cover local IPs if they match.

**Comment by andy-ning** [comment]
Thanks @tobiasbrunner , it makes sense. I believe these local IKE_SAs don't hurt other than having some performance impact on the hosts.
Another question, for the local IKE_SAs in our system, why there are no CHILD_SAs?  My understanding is that CHILD_SAs are the actual IPSec SAs, if there are no CHILD_SAs, the traffic is not really protected.

**Reply by tobiasbrunner** [reply]
No idea, read the logs.

**Comment by andy-ning** [comment]
We are actually integrating strongswan to the StarlingX opensource project (https://www.starlingx.io/) to protect traffics among hosts. A STX based cluster will have multiple hosts, with at least two hosts (controller-0 and controller-1) as the minimum. (This the system I'm testing now, which has only two controller hosts). One of the two controllers will be active, having a unit IP address plus a floating IP address. In my test system:

**controller-0 is active, having two IP addresses:**
9: vlan877@ens786f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP group default qlen 1000
    link/ether b4:96:91:b6:80:38 brd ff:ff:ff:ff:ff:ff
    inet6 **fdff:719a:bf60:2240::2/64** scope global deprecated
       valid_lft forever preferred_lft 0sec
    inet6 **fdff:719a:bf60:2240::3/64** scope global
       valid_lft forever preferred_lft forever

**controller-1 is standby, having only one IP address:**
9: vlan877@ens786f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP group default qlen 1000
    link/ether b4:96:91:b6:81:fc brd ff:ff:ff:ff:ff:ff
    inet6 **fdff:719a:bf60:2240::4/64** scope global
       valid_lft forever preferred_lft forever

**I've update the swanctl.conf on both controllers as following:**
- controller-0:
connections {
        system-nodes {
                reauth_time = 14400
                rekey_time = 3600
                unique = never
                mobike = no
                local_addrs = controller-0, controller
                remote_addrs = %any6
                local {
                        auth = pubkey
                        certs = system-ipsec-certificate-controller-0.crt
                }
                remote {
                        id = CN=*
                        auth = pubkey
                        cacerts = system-local-ca.crt
                }
                children {
                        node {
                                mode = transport
                                start_action = trap
                                local_ts = fdff:719a:bf60:2240::/64
                                remote_ts = fdff:719a:bf60:2240::/64
                        }
                        node_bypass {
                                mode = pass
                                start_action = trap
                                local_ts = fdff:719a:bf60:2240::3, fdff:719a:bf60:2240::2
                                remote_ts = fdff:719a:bf60:2240::3, fdff:719a:bf60:2240::2
                        }
                }
        }
}

- controller-1:
connections {
        system-nodes {
                reauth_time = 14400
                rekey_time = 3600
                unique = never
                mobike = no
                local_addrs = controller-1, controller
                remote_addrs = %any6
                local {
                        auth = pubkey
                        certs = system-ipsec-certificate-controller-1.crt
                }
                remote {
                        id = CN=*
                        auth = pubkey
                        cacerts = system-local-ca.crt
                }
                children {
                        node {
                                mode = transport
                                start_action = trap
                                local_ts = fdff:719a:bf60:2240::/64
                                remote_ts = fdff:719a:bf60:2240::/64
                        }
                        node_bypass {
                                mode = pass
                                start_action = trap
                                local_ts = fdff:719a:bf60:2240::4
                                remote_ts = fdff:719a:bf60:2240::4
                        }
                }
        }
}

I have controller-0 boots up first, with IPSec configured and enabled (charon is running). Then boots/reboots up controller-1. charon will be brought up by systemd during booting. Most of the time, after controller-1 reboots, there will be NO active CHILD_SAs established between the two controllers. But some times there will be active CHILD_SAs established between the controller-0's floating IP and the controller-1's unit IP. I can see charon is struggling to establish SAs between the unit IP addresses. The errors I saw in charon logs are as following (also attached 
[charon_debug_Feb28_C0.log](https://github.com/strongswan/strongswan/files/14435674/charon_debug_Feb28_C0.log)
[charon_debug_Feb28_C1.log](https://github.com/strongswan/strongswan/files/14435677/charon_debug_Feb28_C1.log)
 )

Any idea why there are no SAs established between the two unit IP addresses? (Honestly I failed to see why from the logs as I don't have a deep understanding of IPSec and strongswan). And any suggested changes to the configuration so traffic among hosts can be protected by IPSec? 

- controller-0:
 Wed, 2024-02-28, 14:15:52 10[IKE] <system-nodes|247> giving up after 5 retransmits
Wed, 2024-02-28, 14:15:52 10[IKE] <system-nodes|247> establishing IKE_SA failed, peer not responding
Wed, 2024-02-28, 14:15:52 10[MGR] <system-nodes|247> checkin and destroy IKE_SA system-nodes[247]
Wed, 2024-02-28, 14:15:52 10[IKE] <system-nodes|247> IKE_SA system-nodes[247] state change: CONNECTING => DESTROYING
Wed, 2024-02-28, 14:15:52 10[MGR] checkin and destroy of IKE_SA successful

- controller-1:
Wed, 2024-02-28, 14:15:12 10[IKE] <system-nodes|1106> retransmit 4 of request with message ID 0
Wed, 2024-02-28, 14:15:12 10[NET] <system-nodes|1106> sending packet: from fdff:719a:bf60:2240::3[500] to fdff:719a:bf60:2240::4[500] (710 bytes)
Wed, 2024-02-28, 14:15:12 10[MGR] <system-nodes|1106> checkin IKE_SA system-nodes[1106]
Wed, 2024-02-28, 14:15:12 10[MGR] <system-nodes|1106> checkin of IKE_SA successful
Wed, 2024-02-28, 14:15:12 04[NET] error writing to socket: Invalid argument

Also the swantctl output:
- controller-0:
# swanctl -L
system-nodes: IKEv1/2, reauthentication every 14400s, rekeying every 3600s
  local:  controller-0
  local:  controller
  remote: %any6
  local public key authentication:
    id: CN=ipsec-controller-0
    certs: CN=ipsec-controller-0
  remote public key authentication:
    id: CN=*
    cacerts: CN=starlingx
  node: TRANSPORT, rekeying every 3600s
    local:  fdff:719a:bf60:2240::/64
    remote: fdff:719a:bf60:2240::/64
  node_bypass: PASS, no rekeying
    local:  fdff:719a:bf60:2240::3/128 fdff:719a:bf60:2240::2/128
    remote: fdff:719a:bf60:2240::3/128 fdff:719a:bf60:2240::2/128

# swanctl --list-sa
system-nodes: #262, CONNECTING, IKEv2, 5a9d2857be5e4811_i* 0000000000000000_r
  local  '%any' @ fdff:719a:bf60:2240::3[500]
  remote '%any' @ fdff:719a:bf60:2240::4[500]
  active:  IKE_VENDOR IKE_INIT IKE_NATD IKE_CERT_PRE IKE_AUTH IKE_CERT_POST IKE_CONFIG CHILD_CREATE IKE_AUTH_LIFETIME
system-nodes: #261, ESTABLISHED, IKEv2, 76b143cf1f48774a_i bfc199e1403350e2_r*
  local  'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::2[500]
  remote 'CN=ipsec-controller-1' @ fdff:719a:bf60:2240::4[500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_AES128_XCBC/MODP_3072
  established 232s ago, rekeying in 2228s, reauth in 13713s
  node: #16, reqid 2, INSTALLED, TRANSPORT, ESP:AES_GCM_16-128
    installed 232s ago, rekeying in 3039s, expires in 3728s
    in  c29737f0,  54285 bytes,   532 packets,     0s ago
    out c18ba9e9,  54820 bytes,   450 packets,   207s ago
    local  fdff:719a:bf60:2240::2/128
    remote fdff:719a:bf60:2240::4/128

- controller-1:
# swanctl -L
system-nodes: IKEv1/2, reauthentication every 14400s, rekeying every 3600s
  local:  controller-1
  local:  controller
  remote: %any6
  local public key authentication:
    id: CN=ipsec-controller-1
    certs: CN=ipsec-controller-1
  remote public key authentication:
    id: CN=*
    cacerts: CN=starlingx
  node: TRANSPORT, rekeying every 3600s
    local:  fdff:719a:bf60:2240::/64
    remote: fdff:719a:bf60:2240::/64
  node_bypass: PASS, no rekeying
    local:  fdff:719a:bf60:2240::4/128
    remote: fdff:719a:bf60:2240::4/128

# swanctl --list-sa
(unnamed): #1173, CONNECTING, IKEv2, 5a9d2857be5e4811_i 672593f95b20f912_r*
  local  '%any' @ fdff:719a:bf60:2240::4[500]
  remote '%any' @ fdff:719a:bf60:2240::3[500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_AES128_XCBC/MODP_3072
  passive: IKE_CERT_PRE IKE_AUTH IKE_CERT_POST IKE_CONFIG CHILD_CREATE IKE_AUTH_LIFETIME IKE_MOBIKE
system-nodes: #1171, CONNECTING, IKEv2, 2940248e0bb359bf_i* 0000000000000000_r
  local  '%any' @ fdff:719a:bf60:2240::3[500]
  remote '%any' @ fdff:719a:bf60:2240::4[500]
  active:  IKE_VENDOR IKE_INIT IKE_NATD IKE_CERT_PRE IKE_AUTH IKE_CERT_POST IKE_CONFIG CHILD_CREATE IKE_AUTH_LIFETIME
system-nodes: #1170, CONNECTING, IKEv2, 5bd6668ef1b9c3bb_i* 0000000000000000_r
  local  '%any' @ fdff:719a:bf60:2240::4[500]
  remote '%any' @ fdff:719a:bf60:2240::3[500]
  active:  IKE_VENDOR IKE_INIT IKE_NATD IKE_CERT_PRE IKE_AUTH IKE_CERT_POST IKE_CONFIG CHILD_CREATE IKE_AUTH_LIFETIME
system-nodes: #1169, ESTABLISHED, IKEv2, 76b143cf1f48774a_i* bfc199e1403350e2_r
  local  'CN=ipsec-controller-1' @ fdff:719a:bf60:2240::4[500]
  remote 'CN=ipsec-controller-0' @ fdff:719a:bf60:2240::2[500]
  AES_CBC-128/HMAC_SHA2_256_128/PRF_AES128_XCBC/MODP_3072
  established 270s ago, rekeying in 1986s, reauth in 12235s
  node: #18, reqid 1, INSTALLED, TRANSPORT, ESP:AES_GCM_16-128
    installed 270s ago, rekeying in 3055s, expires in 3690s
    in  c18ba9e9,  67020 bytes,   528 packets,     2s ago
    out c29737f0,  62303 bytes,   622 packets,     3s ago
    local  fdff:719a:bf60:2240::4/128
    remote fdff:719a:bf60:2240::2/128

**Reply by tobiasbrunner** [reply]
First, I'd recommend you configure the bypass policies in separate connection sections with `remote_addrs` set to `::1` as described [in the docs](https://docs.strongswan.org/docs/5.9/config/quickstart.html#_passthroughbypass_policies).

> Any idea why there are no SAs established between the two unit IP addresses

What exactly does that mean? What are the "unit" IP addresses? Note that transport mode SAs can only be established between the IPs the peers actually use to communicate (i.e. they have to be the same as the IKE endpoints). These addresses in turn are determined by the routing table(s).

**Comment by andy-ning** [comment]
> What exactly does that mean? What are the "unit" IP addresses?

Each of the two controllers has a fixed IP addresses (what we call the unit IP). At any moment of time, one of the controllers is active and get an extra IP addresses (what we call the floating IP). In my test, controller-0 is active, so it has a fixed unit IP (**fdff:719a:bf60:2240::3**) plus a floating IP (**fdff:719a:bf60:2240::2**), while controller-1 has only its fixed unit IP (**fdff:719a:bf60:2240::4**). There are traffics:
controller-0 unit IP  <-> controller-1 unit IP
controller-0 floating IP <-> controller-1 unit IP

**Comment by andy-ning** [comment]
I'm very confused by the **\<conn\>.local_addrs|remote_addrs and \<conn\>.\<child\>.local_ts|remote_ts**. My understanding is that the IPSec policies eventually installed in kernel are based on \<conn\>.\<child\>.local_ts|remote_ts. And \<conn\>.local_addrs|remote_addrs are used by charon to do the IKE SA negotiation. Once charon finishes the IKE negotiation, it will generate policies based on \<conn\>.\<child\>.local_ts|remote_ts and the negotiation results, then install them in kernel.

Once the policies are installed, IP packets that match these policies will be handled by IPsec accordingly. 

Are there any documents that explain the relationship between **\<conn\>.local_addrs|remote_addrs and \<conn\>.\<child\>.locat_ts|remote_ts**? And how are they used to generate the IPSec policies?

**Reply by tobiasbrunner** [reply]
Your understanding is basically correct. However, you have to consider the IPsec mode. With tunnel mode, which encapsulates the complete IP packet in an ESP packet that gets a new IP header, the traffic selectors can be different than the endpoint addresses used for the IKE and IPsec SAs. However, with transport mode that's not the case as the original IP header is used and only the payload is encapsulated in ESP. So this mode can only protect traffic where the addresses and traffic selectors match.

There is another mode called BEET (Bound End-to-End Tunnel), which is something in between. Like transport mode it doesn't add an additional IP header but at the endpoints the IPs in the header are mapped to different ones (one to one mapping). It's not very common and was never published as an RFC but strongSwan supports it. In this case the addresses in `local|remote_addrs` can be different than the `/128` traffic selectors in `local|remote_ts`.

Note that the wildcard trap policies (where you don't specify the addresses, only the traffic selectors) can only be used with transport mode.

**Comment by andy-ning** [comment]
Thanks again for the explanation. 
I feel I need a deeper understanding on how strongswan IPSec is working, so I drew a simple diagram of a IP packet flowing through the processing. Do you mind correcting me if anything is incorrect?  
[IPSec packet flow.pptx](https://github.com/strongswan/strongswan/files/14466550/IPSec.packet.flow.pptx)

**Reply by tobiasbrunner** [reply]
Looks about right. Step 6 depends on the original config, that is, whether a remote address was actually configured or not. If not, then the remote address from the matching packet will be used. And similar for step one, if a remote address is configured, there is no point in configuring `remote_ts` when using transport mode as the default (`dynamic`) gets resolved to that IP anyway.

Also, the second sentence isn't entirely correct as there might be rekeyings of IPsec and IKE SAs, maybe also DPDs, or even MOBIKE updates.

By the way, IPsec is the correct capitalization, not IPSec.

**Reply by andy-ning** [reply]
Thanks for the correction.

What we are facing right now is that, controller-0 and controller-1 are failing to establish IKE SAs. We monitor traffic on udp port 500 and 4500 using tcpdump on both hosts. We can see on one host packets coming in from the other host, but don't see packets going out to the other host. This seems to cause the IKE negotiation fail.

Just restarting charon doesn't make IKE SA established. But if we stop charon on both hosts, the run "ip xfrm policy flush" and "ip xfrm stat flush" on both hosts, then start charon again, IKE SAs and CHILD SAs will be established successfully. (We have to do this on both hosts to make SAs established)

**Reply by tobiasbrunner** [reply]
Can't help you with that without configs, logs etc.

**Reply by andy-ning** [reply]
The swanctl.conf and charon log for controller-0 and controller-1 are here: 
[Mar07_2024.zip](https://github.com/strongswan/strongswan/files/14528648/Mar07_2024.zip)

This is the most conservative config, ie, I use specific IP addresses instead of subnets.

Again this is a two hosts system (controller-0 and controller-1). **controller-0 has two IP addresses:  fdff:719a:bf60:2080::2 and fdff:719a:bf60:2080::3, controller-1 has one IP address: fdff:719a:bf60:2080::4**

The logs were captured when controller-0 was up and running, and controller-1 was rebooted to running state.

Please let me know if any further information can help.

**Reply by tobiasbrunner** [reply]
> This is the most conservative config, ie, I use specific IP addresses instead of subnets.

For the traffic selectors, not for `remote_addrs`. So this is not a conservative config, it's still using wildcard trap policies (just limited to single IPs).

One problem with your config is on controller-1. It has two configs that are essentially the same. They only differ in the traffic selectors for the child configs. This won't work as a responder because the config is selected based on the IP addresses (matching `local|remote_addrs`) and the identities. Since those are equal for both configs, the first one is selected. And if controller-0 connects from the floating IP that won't match the traffic selector of that config.

However, you don't get that far as there also seems to be some other issue. Packets are seemingly getting dropped somewhere, or not received for some other reason. One possibility is that two IKE daemons are running, e.g. charon and charon-systemd. If that's not the case try to capture traffic to see where those IKE_SA_INIT messages are lost.

**Reply by andy-ning** [reply]
That's right, we are seeing one side's response packets never come back to the sender's side. But what's interesting is, if we stop ipsec on both sides, then "ip xfrm policy flush" and "ip xfrm state flush" on both sides, then start ipsec. After this, the IKE SA and CHILD_SA can established. It just doesn't work when controller-1 is rebooted.

**Reply by andy-ning** [reply]
FYI, I think we hit this: https://wiki.strongswan.org/projects/strongswan/wiki/IPv6NDP/1
By bypassing ICMPv6 in config, SAs are established after reboot.
